File: backend/lib/repousse_web/schemas/project.ex
defmodule RepousseWeb.Schemas.Project do
  @moduledoc false
  require OpenApiSpex
  alias OpenApiSpex.Schema

  OpenApiSpex.schema(
    %{
      title: "Project",
      type: :object,
      properties: %{
        id: %Schema{type: :string, format: :uuid},
        name: %Schema{type: :string},
        description: %Schema{type: :string, nullable: true},
        management_type: %Schema{type: :string, enum: ["individual", "collective"]},
        address: %Schema{type: :string, nullable: true},
        lat: %Schema{type: :number, format: :float, nullable: true},
        lng: %Schema{type: :number, format: :float, nullable: true},
        surface_m2: %Schema{type: :number, format: :float, nullable: true},
        soil_type: %Schema{type: :string, nullable: true},
        publication_status: %Schema{type: :string, enum: ["private", "public", "unpublished"]},
        published_at: %Schema{type: :string, format: "date-time", nullable: true},
        archived_at: %Schema{type: :string, format: "date-time", nullable: true},
        owner_id: %Schema{type: :string, format: :uuid, nullable: true},
        cover_image_url: %Schema{type: :string, nullable: true},
        inserted_at: %Schema{type: :string, format: "date-time"},
        updated_at: %Schema{type: :string, format: "date-time"}
      },
      required: [:id, :name],
      example: %{
        "id" => "c1b2c3d4-0000-4000-8000-000000000000",
        "name" => "Verger partagé du Clos",
        "description" => "Un projet de plantation collectif",
        "management_type" => "collective",
        "address" => nil,
        "lat" => nil,
        "lng" => nil,
        "surface_m2" => nil,
        "soil_type" => nil,
        "publication_status" => "private",
        "published_at" => nil,
        "archived_at" => nil,
        "owner_id" => "a1b2c3d4-0000-4000-8000-000000000000",
        "cover_image_url" => nil,
        "inserted_at" => "2026-01-15T10:00:00Z",
        "updated_at" => "2026-01-15T10:00:00Z"
      }
    },
    struct?: false
  )
end

File: webapp/src/types/project.ts
export type ManagementType = "individual" | "collective";

export type PublicationStatus = "private" | "public" | "unpublished";

export interface ProjectPreferredSpecies {
  id: string;
  project_id: string;
  taxon_id: string;
  inserted_at: string;
  updated_at: string;
}

export interface Project {
  id: string;
  name: string;
  description: string | null;
  management_type: ManagementType;
  address: string | null;
  lat: number | null;
  lng: number | null;
  surface_m2: number | null;
  soil_type: string | null;
  publication_status: PublicationStatus;
  published_at: string | null;
  archived_at: string | null;
  owner_id: string | null;
  cover_image_url: string | null;
  preferred_species: ProjectPreferredSpecies[];
  inserted_at: string;
  updated_at: string;
}

export const MANAGEMENT_TYPES: ManagementType[] = ["individual", "collective"];

export const MANAGEMENT_TYPE_LABELS: Record<ManagementType, string> = {
  individual: "Individuelle",
  collective: "Collective",
};

export const PUBLICATION_STATUSES: PublicationStatus[] = ["private", "public", "unpublished"];

export const PUBLICATION_STATUS_LABELS: Record<PublicationStatus, string> = {
  private: "Privé",
  public: "Public",
  unpublished: "Dépublié",
};

export const PUBLICATION_STATUS_COLORS: Record<PublicationStatus, string> = {
  private: "bg-slate-100 text-slate-800 dark:bg-slate-800/50 dark:text-slate-300",
  public: "bg-green-100 text-green-800 dark:bg-green-900/30 dark:text-green-300",
  unpublished: "bg-red-100 text-red-800 dark:bg-red-900/30 dark:text-red-300",
};

/** Allowed forward/backward transitions from a given status (moderation-only "unpublished" not exposed here). */
export const PUBLICATION_STATUS_TRANSITIONS: Record<PublicationStatus, PublicationStatus[]> = {
  private: ["public"],
  public: ["private"],
  unpublished: [],
};

export function findProjectById(id: string, rows: Project[]): Project | null {
  return rows.find((r) => r.id === id) ?? null;
}
